pkg/chartutil/values.go

```diff
@@ -326,25 +275,50 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
 func coalesceTables(dst, src map[string]interface{}, chartName string) map[string]interface{} {
 	// Because dest has higher precedence than src, dest values override src
 	// values.
+
+	rv := make(map[string]interface{})
 	for key, val := range src {
-		if istable(val) {
-			if innerdst, ok := dst[key]; !ok {
-				dst[key] = val
-			} else if istable(innerdst) {
-				coalesceTables(innerdst.(map[string]interface{}), val.(map[string]interface{}), chartName)
-			} else {
-				log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
-			}
+		dv, ok := dst[key]
+		if !ok { // if not in dst, then copy from src
+			rv[key] = val
 			continue
-		} else if dv, ok := dst[key]; ok && istable(dv) {
-			log.Printf("Warning: Merging destination map for chart '%s'. The destination item '%s' is a table and ignoring the source '%s' as it has a non-table value of: %v", chartName, key, key, val)
+		}
+		if dv == nil { // if set to nil in dst, then ignore
+			// When the YAML value is null, we skip the value's key.
+			// This allows Helm's various sources of values (value files or --set) to
+			// remove incompatible keys from any previous chart, file, or set values.
 			continue
-		} else if !ok { // <- ok is still in scope from preceding conditional.
-			dst[key] = val
+		}
+
+		srcTable, srcIsTable := val.(map[string]interface{})
+		dstTable, dstIsTable := dv.(map[string]interface{})
+		switch {
+		case srcIsTable && dstIsTable: // both tables, we coalesce
+			rv[key] = coalesceTables(dstTable, srcTable, chartName)
+		case srcIsTable && !dstIsTable:
+			log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
+			// despite message in warning, we appear to do exactly that, and do take the dst value
```

Comment:
@aeijdenberg Yes this is the only thing I noticed here. Somehow, when nullifying a key this warning prints. Example from community helm/charts:
```console
$ helm install stable/drupal --set livenessProbe.httpGet=null --set livenessProbe.exec.command[0]='php -v' --debug --dry-run
...
2019/04/23 14:00:41 Warning: Merging destination map for chart 'drupal'. Cannot overwrite table item 'httpGet', with non table value: map[path:/user/login port:http]
```

I agree with your comment - this warning is confusing. It strangely seems to imply we're attempting to overwrite the original key with its value.

I wonder though, how this message ever triggers for a nullified key, since we `continue` above in that case?

@hickeyma since you wrote #4974 would you be willing to take a look at that side of these code changes?

Reply:
#4974 does not change any existing logic. It adds chart name to existing warning/error messages with some modifications to the message.

From looking at the code, this message is generated from [here](https://github.com/helm/helm/blob/master/pkg/chartutil/values.go#L336)
 
It is part of the if/else ladder:
```
if innerdst, ok := dst[key]; !ok {
	dst[key] = val
} else if istable(innerdst) {
	coalesceTables(innerdst.(map[string]interface{}), val.(map[string]interface{}), chartName)
} else {
	log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
}
continue
```
It seems to check that the destination map is valid yaml table and if is not, it prints the warning message. It continues after the ladder.

Reply:
Sorry, was looking at previous version of the file and not the updated one.

From my reading of the code (and not known the background to it), it would seem that you could have YAML returned from `dv, ok := dst[key]` which would not be null but would not match a valid YAML table and therefore be caught here: `case srcIsTable && !dstIsTable:` and so output the warning you mentioned.

Reply:
Never mind, somehow I must have made a mistake in debugging. I don't see the error now 😅 Resolving this one!